pkg/platform/runtime/buildplan/buildplan.go

```diff
@@ -291,3 +291,132 @@ func NewNamedMapFromBuildPlan(build *model.Build) (artifact.NamedMap, error) {
 
 	return res, nil
 }
+
+// BuildtimeArtifacts iterates through all artifacts in a given build and
+// adds the artifact's dependencies to a map. This is different from the
+// runtime dependency calculation as it includes ALL of the input artifacts of the
+// step that generated each artifact. This is useful for when we are using the
+// BuildLogStreamer as it operates on the older recipeID and will include
+// more artifacts than what we originally calculated in the runtime closure.
+func BuildtimeArtifacts(build *model.Build) (artifact.Map, error) {
+	result := make(artifact.Map)
+	lookup := make(map[strfmt.UUID]interface{})
+
+	for _, artifact := range build.Artifacts {
+		lookup[artifact.NodeID] = artifact
+	}
+	for _, step := range build.Steps {
+		lookup[step.StepID] = step
+	}
+	for _, source := range build.Sources {
+		lookup[source.NodeID] = source
+	}
+
+	for _, a := range build.Artifacts {
+		// Since we are using the BuildLogStreamer, we need to add all of the
+		// artifacts that have been submitted to be built except for the artifacts
+		// of the builder mime type as the BuildLogStreamer does not send events
+		// for builder artifacts.
+		if a.MimeType == model.XActiveStateBuilderMimeType {
+			continue
+		}
+
+		_, ok := result[strfmt.UUID(a.NodeID)]
+		if !ok && a.Status != model.ArtifactNotSubmitted {
+			deps := make(map[strfmt.UUID]struct{})
+			for _, depID := range a.RuntimeDependencies {
+				artifact, ok := lookup[depID].(*model.Artifact)
+				if ok {
+					if artifact.MimeType == model.XActiveStateBuilderMimeType {
+						logging.Debug("Skipping active state builder artifact: %s", depID)
+						continue
+					}
+				}
+
+				deps[depID] = struct{}{}
+				recursiveDeps, err := generateBuildtimeDependencies(depID, lookup, deps)
+				if err != nil {
+					return nil, errs.Wrap(err, "Could not resolve runtime dependencies for artifact: %s", depID)
+				}
+				for id := range recursiveDeps {
+					deps[id] = struct{}{}
+				}
+			}
+
+			var uniqueDeps []strfmt.UUID
+			for id := range deps {
+				if _, ok := deps[id]; !ok {
+					continue
+				}
+				uniqueDeps = append(uniqueDeps, id)
+			}
+
+			info, err := getSourceInfo(a.GeneratedBy, lookup)
+			if err != nil {
+				return nil, errs.Wrap(err, "Could not resolve source information")
+			}
```

Comment:
Please add comments explaining what this is doing.

Reply:
That's what the function docstring is doing.

Reply:
I meant the individual code. I didn't want to flag every piece of code, just saying that this whole chunk has a lot of logic and I'd like to be able to ready over it quickly by reading a short comment before each "block" explaining what this code is doing so that I can just read the code with context rather than read the code and trying to parse context.